### PR TITLE
Reduce duplication pushing posts.

### LINKED ISSUE #997: Reduce duplication pushing post to connections
### Is your enhancement related to a problem? Please describe.

With the introduction of a post abstraction much of the duplication in the External and Network connection classes for pushing a post can be removed.

Follow up to PR #964 for issue #207 

### Designs

_No response_

### Describe alternatives you've considered

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by peterwilsoncc**
### Description of the Change

Replaces duplicate code in the network and external connection classes by using the post abstraction.

Closes #997 

### How to test the Change

This requires testing on both **network (internal)** connections and **external** connections. The steps for each are the same:

* Create a post on the source site, test with
  * terms
  * content
  * media
  * post meta (this mini [plugin updates an item on each save](https://gist.github.com/peterwilsoncc/61eeb5f2337cdf4976608a5123acbb1d))
* ensure that push as draft/push as published feature works as expected
* ensure the linking, unlinking of posts works as expected (including updates while unlinked)
* ensure time stamp updates are unchanged from the develop branch (including modified times, this may require looking in the database)
* using xdebug to compare the payloads from develop and this branch would be lovely too.

### Changelog Entry

> Changed - Reduced duplicate code pushing posts to connections.

### Credits

Props @peterwilsoncc, @cadic.

### Checklist:

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [x] All new and existing tests pass.

### CODE REVIEW

File: includes/classes/InternalConnections/NetworkSiteConnection.php [hunk 1]
```diff
@@ -58,160 +59,149 @@ public function __construct( WP_Site $site ) {
...
+	 *
+	 *     @type int    $remote_post_id Post ID on remote site. If not provided,
+	 *                                  a new post will be created.
+	 *     @type string $post_status    The post status to use on the remote site.
+	 *                                  Ignored when updating posts. Default 'publish'.
+	 * }
 	 * @since  0.8
 	 * @return array|\WP_Error
 	 */
 	public function push( $post_id, $args = array() ) {
```
Comment (cadic) [inline includes/classes/InternalConnections/NetworkSiteConnection.php]: Within the current PR, in [WordPressExternalConnection::push()](https://github.com/10up/distributor/pull/999/files#diff-fb0e5652598ac73446e096480da914a26fa59b2e929ecb314b4169ac51ff1d53R604), you've updated the first argument to be `int|WP_Post`. I wonder should we do the same here for consistency?

File: tests/php/DistributorPostTest.php [hunk 2]
```diff
@@ -45,6 +45,8 @@ public function setup_common() {
 		\WP_Mock::userFunction( '_prime_post_caches' );
 		\WP_Mock::userFunction( 'update_object_term_cache' );
 		\WP_Mock::userFunction( 'update_postmeta_cache' );
+		\WP_Mock::userFunction( 'switch_to_blog' );
```
Comment (cadic) [inline tests/php/DistributorPostTest.php:48]: Looks like it's safe to remove the same user functions from lines 166-167 now after they are declared in the setup_common()?

**Review by cadic — CHANGES_REQUESTED** [review]
I've tested this PR and some workflows:

Created a new post on source site with images, categories, tags and custom meta.

|Workflow|Network|External conn|
|-|-|-|
| Push as Draft | ✅ | ✅ |
| Push as Published | ✅ | ✅ |
| Change the original (see changes reflected to copies) | ✅ | ✅ |
| Unlink from the original, make changes, re-link back, see the copy reset to original| ✅ | ✅ |

Works great!

The code looks good, I just have a couple suggestions about the argument type and removing the duplicate user functions from Unit test.

File: includes/classes/InternalConnections/NetworkSiteConnection.php (hunk 1 above)
Comment (peterwilsoncc) [reply includes/classes/InternalConnections/NetworkSiteConnection.php]: Excellent idea, pushed in 4b37f89b6fc5b441af315c18080a7b9f0677d984

File: tests/php/DistributorPostTest.php (hunk 2 above)
Comment (peterwilsoncc) [reply tests/php/DistributorPostTest.php:48]: Removed in 6d61fb402d50cc7ba078a0a384f31c698d90643b

**Comment by peterwilsoncc** [comment]
Thanks @cadic -- I've followed up each of your suggestions.

**Review by cadic — APPROVED** [review]

**Pull request merged**